src/cell_group.hpp

```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
```

Comment:
@w-klijn is right, in that this function is a bit hard to follow. Sometimes a longer function can make complicated code easier to understand, but in this case things like the sampling logic could be refactored.

But this is a work in progress, it should be cleaner once much of the event and sampling logic has been moved to the back end.